python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py

```diff
@@ -159,3 +179,143 @@ def ensure_freshness_checks(checks: Sequence[AssetChecksDefinition]) -> None:
                 ),
                 f"Asset check {check_spec.key} didn't have expected metadata. Please ensure that the asset check is a freshness check.",
             )
+=======
+def get_last_update_time_lower_bound(
+    freshness_cron: Optional[str],
+    current_time: datetime.datetime,
+    lower_bound_delta: datetime.timedelta,
+    freshness_cron_timezone: str,
+) -> datetime.datetime:
+    """Get the lower bound of the last_update_time for the given cron schedule."""
+    if freshness_cron:
+        return (
+            check.not_none(
+                get_latest_completed_cron_tick(
+                    freshness_cron=freshness_cron,
+                    current_time=current_time,
+                    timezone=freshness_cron_timezone,
+                )
+            )
+            - lower_bound_delta
+        )
+    else:
+        return current_time - lower_bound_delta
+
+
+def get_latest_complete_partition_key(
+    freshness_cron: Optional[str],
+    current_time: datetime.datetime,
+    freshness_cron_timezone: str,
+    partitions_def: Optional[TimeWindowPartitionsDefinition],
+) -> Optional[str]:
+    """Get the latest complete partition key for the given cron schedule."""
+    if not partitions_def:
+        return None
+    latest_cron_tick_time = check.not_none(
+        get_latest_completed_cron_tick(
+            freshness_cron=freshness_cron,
+            current_time=current_time,
+            timezone=freshness_cron_timezone,
+        )
+    )
+    time_window = check.not_none(partitions_def.get_prev_partition_window(latest_cron_tick_time))
+    return partitions_def.get_partition_key_range_for_time_window(time_window).start
+
+
+def build_freshness_check(
+    asset_key: AssetKey,
+    freshness_cron: Optional[str],
+    freshness_cron_timezone: str,
+    severity: AssetCheckSeverity,
+    lower_bound_delta: datetime.timedelta,
+    asset_property_enforcement_lambda: Optional[Callable[[AssetsDefinition], bool]],
+) -> AssetChecksDefinition:
+    @asset_check(
+        asset=asset_key,
+        description="Evaluates freshness for targeted asset.",
+        name="freshness_check",
+    )
+    def the_check(context: AssetExecutionContext) -> AssetCheckResult:
+        if asset_property_enforcement_lambda:
+            asset_property_enforcement_lambda(
+                context.job_def.asset_layer.asset_graph.get(asset_key).assets_def
+            )
+        current_time = pendulum.now(
+            tz=freshness_cron_timezone
+        )  # Will be replaced with "effective_timestamp" once it's available.
```

Comment:
What's "effective_timestamp" and where would it come from?

Reply:
Something nick discussed earlier, about threading a single "effective timestamp" through the entire execution stack instead of continuously calling pendulum.now()

Reply:
`effective_dt` was the terminology that was agreed upon in the `AssetGraphView` [PR](https://github.com/dagster-io/dagster/pull/20312).

Reply:
Got it - makes sense.

That said, generally not a fan of putting TODOs in comments, because plans / names can change and easy for comments to get stale.

Reply:
Agreed. Will remove before we land, just thought it was helpful from a review perspective